keps/sig-network/0031-20181017-kube-proxy-services-optional.md

```diff
@@ -0,0 +1,135 @@
+
+#### Story 1
+
+As a cluster operator, operating a cluster using a service mesh I want to be able to disable the kube-proxy service implementation for services in that mesh to reduce overall load on the whole cluster
+
+### Implementation Details/Notes/Constraints
+
+#### Overview
+
+It is important for overall scalability that kube-proxy does not watch for service/endpoint changes that it is not going to affect. This can save a lot of load on the apiserver, networking, and kube-proxy itself by never requesting the updates in the first place. As such, annotating the services directly is considered insufficient as the kube-proxy would still have to watch for changed to the service.
+
+The proposal is to make this feature available at the namespace level. We will support a new label for namespaces: `networking.k8s.io/service-proxy=disabled`
+
+When this label is set, kube-proxy will behave as if services in that namespace do not exist. None of the functionality that kube-proxy provides will be available for services in that namespace.
+
+It is expected that this feature will mainly be used on large clusters with lots (>1000) of services. Any use of this feature in a smaller cluster will have negligible impact.
+
+The envisioned cluster that will make use of this feature looks something like the following:
+* Most/all traffic from outside the cluster is handled by gateways, such that each service in the cluster does not need a nodePort
+* These small number of entry points into the cluster are a part of the service mesh
+* There are many micro-services in the cluster, all a part of the service mesh, that are only accessed from inside the service mesh
+  * These services are in a separate namespace from the gateways
+
+#### Design
+
+Currently, when ProxyServer starts up it creates informers for all Service (ServiceConfig) and Endpoints (EndpointsConfig) objects using a single shared informer factory. The new design will make these previous objects be per-namespace, and only listen on namespaces that are not 'disabled'.
+
+The ProxyServer type will be updated with the following new methods:
+* func (s *ProxyServer) StartWatchingNamespace(ns string)
+  * Check if namespace is currently watched, if it is then return
+  * Create a shared informer factory configured with the namespace
+  * Create a ServiceConfig and EndpointsConfig object using the shared informer factory
+* func (s *ProxyServer) StopWatchingNamespace(ns string)
+  * Check if namespace is currently watched, if it is not then return
+  * Stop the ServiceConfig and EndpointsConfig for that namespace
+  * Send deletion events for all objects those configs knew about
+  * Delete the config objects
+
+At startup time, ProxyServer will create an informer for all Namespace objects.
+* When a namespace objects is created or updated:
+  * Check for the above label, and if it is not set or is not 'disabled':
+    * StartWatchingNamespace()
+  * Else:
+    * StopWatchingNamespace()
```

Comment:
This turns the number of watches on the master from O(N) to O(M*N).  If M is much greater than 5-10 this would be “bad” for medium or large clusters (watches aren’t that cheap).

Is this a solution only for small M or N?

Reply:
Maybe a better solution is watching with label selector? e.g. `proxy.kubernetes.io/disabled!=true`

Reply:
Thank you @feiskyer, that sounds like a great suggestion.

Reply:
AFAIK there's no way to select namespaces to watch into.  That would be neat, but I don't think it exists.